### PR TITLE
ENH PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/10250

### CODE REVIEW

File: src/Extension/FluentDirectorExtension.php [hunk 1]
```diff
@@ -166,9 +166,9 @@ protected function getRuleController($existingRule, $localeObj)
      */
     protected function insertRuleBefore(array $rules, $key, array $rule, $prependIfMissing = true)
     {
-        $i = array_search($key, array_keys($rules));
+        $i = array_search($key, array_keys($rules ?? []));
```
Comment (tractorcow) [inline src/Extension/FluentDirectorExtension.php:169]: I assume `array` param type precludes null values. Is that not true in 8.1 anymore?

Comment (tractorcow) [reply src/Extension/FluentDirectorExtension.php:169]: ok,  so it's more to do with changes to internal function arguments than typehints, apparently. I think in this case we can rely on the typehint. :)

**Comment by tractorcow** [comment]
In general, not harmless, but lots of unnecessary nullable checks. I can approve if you are ready.

**Comment by emteknetnz** [comment]
PHP 8.1 throws deprecation warnings if you pass null into non-nullable native functions now.  A whole lot of fun.

Solution was a code writer which does create lots of unnecessary nullable checks.  We've done the same on framework and all other supported modules

Happy to have approved, though I'd like to keep draft for now and unmerged on the chance that there's something that was missed

**Comment by emteknetnz** [comment]
@tractorcow This is ready for review now. Thanks.

**Comment by GuySartorelli** [comment]
@tractorcow there's some discussion in silverstripe/silverstripe-admin#1294 about the general approach which may be useful context.

**Review by tractorcow — APPROVED** [review]

**Comment by emteknetnz** [comment]
@tractorcow could you please tag 4.7.0, we'll need this to ensure the next CMS release is PHP 8.1 compatible. Thanks

**Pull request merged**